File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
   // topological ordering - in bfs using in degree and in dfs using stacks
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
      
        vector<int> inDegree(numCourses,0);
        vector<int> ans;
        vector<vector<int>> adj(numCourses);
        for(int i=0;i<prerequisites.size();i++){
            inDegree[prerequisites[i][0]]++;
            adj[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        queue<int> q;
        for(int i=0;i<numCourses;i++){
           // cout<<inDegree[i]<<endl;
            if(inDegree[i]==0)q.push(i);
        }
        //cout<<"ggf"<<endl;
        int count = 0;
        while(!q.empty()){
            count++;
            int node = q.front();
            ans.push_back(node);
           // cout<<node<<" "<<count<<endl;
            q.pop();
            for(int i : adj[node]){
                inDegree[i]--;
                if(inDegree[i]==0)q.push(i);
            }
        }
        //cout<<count<<endl;
      if (count==numCourses) return ans;
      return {};
        
      
    }
};

File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int n = numCourses;
        List<Integer>[] adj = new ArrayList[n];
        Stack<Integer> s = new Stack<>();
        int[] visited = new int[n];
        int[] pathVisited = new int[n];
        for(int i=0;i<n;i++) adj[i]= new ArrayList<>();
        for(int[] temp : prerequisites)
        {
            adj[temp[1]].add(temp[0]);
        }
        for( int i =0 ;i<n;i++){
            if(visited[i]==0)
            {
                if(DFS(i,adj,s,visited,pathVisited)==false) return new int[0];
            }
        }
        int[] ans = new int[n];
        int j =0;
        while(!s.isEmpty()) ans[j++]= s.pop();
         return ans;
    }
    public boolean DFS(int node, List<Integer>[] adj, Stack<Integer> s, int[] visited,int[] pathVisited)
    {
        visited[node] = 1;
        pathVisited[node] = 1;
        for(int i : adj[node]){
            if(visited[i]==0) {
                if(DFS(i,adj,s,visited,pathVisited)==false) return false;
            }
            else{
                if(visited[i]==pathVisited[i]) return false;
            }
        }
        pathVisited[node] =0;
        s.push(node);
        return true;
        }
}